File: sdk/python/tests/test_engine.py
"""Tests for engine module."""

import pytest
from substrait_compliance import (
    EngineInfo,
    EngineCapabilities,
    ComplianceEngine,
    ComplianceResult,
    TestStatus
)


def test_engine_info():
    """Test EngineInfo creation."""
    info = EngineInfo(
        name="TestEngine",
        version="1.0.0",
        vendor="TestVendor",
        description="Test engine"
    )
    
    assert info.name == "TestEngine"
    assert info.version == "1.0.0"
    assert str(info) == "TestEngine 1.0.0 by TestVendor"


def test_engine_capabilities():
    """Test EngineCapabilities."""
    caps = EngineCapabilities(
        supported_relations=["read", "filter"],
        supported_functions=["add", "subtract"]
    )
    
    assert caps.supports_relation("read")
    assert caps.supports_relation("filter")
    assert not caps.supports_relation("join")
    
    assert caps.supports_function("add")
    assert not caps.supports_function("multiply")


class MockEngine(ComplianceEngine):
    """Mock engine for testing."""
    
    def get_info(self):
        return EngineInfo("Mock", "1.0", "Test")
    
    def get_capabilities(self):
        return EngineCapabilities()
    
    def execute_plan(self, plan_bytes, input_data):
        return ComplianceResult("test", TestStatus.PASSED)
    
    def validate_plan(self, plan_bytes):
        return ComplianceResult("test", TestStatus.PASSED)


def test_mock_engine():
    """Test mock engine implementation."""
    engine = MockEngine()
    
    info = engine.get_info()
    assert info.name == "Mock"
    
    result = engine.execute_plan(b"", {})
    assert result.is_success()

File: sdk/cpp/tests/test_engine.cpp
#include <gtest/gtest.h>
#include <substrait_compliance/engine.h>

using namespace substrait::compliance;

class MockEngine : public ComplianceEngine {
public:
    EngineInfo get_info() const override {
        return EngineInfo("MockEngine", "1.0.0", "Test");
    }
    
    EngineCapabilities get_capabilities() const override {
        EngineCapabilities caps;
        caps.add_relation("read").add_function("add");
        return caps;
    }
    
    ComplianceResult execute_plan(
        const std::vector<uint8_t>&,
        const TableCollection&
    ) override {
        return ComplianceResult("test", TestStatus::PASSED);
    }
    
    ComplianceResult validate_plan(const std::vector<uint8_t>&) override {
        return ComplianceResult("validation", TestStatus::PASSED);
    }
};

TEST(EngineInfoTest, Construction) {
    EngineInfo info("TestEngine", "2.0.0", "TestCorp");
    
    EXPECT_EQ(info.name, "TestEngine");
    EXPECT_EQ(info.version, "2.0.0");
    EXPECT_EQ(info.vendor, "TestCorp");
}

TEST(EngineCapabilitiesTest, AddRelations) {
    EngineCapabilities caps;
    caps.add_relation("read").add_relation("filter");
    
    EXPECT_TRUE(caps.supports_relation("read"));
    EXPECT_TRUE(caps.supports_relation("filter"));
    EXPECT_FALSE(caps.supports_relation("join"));
}

TEST(EngineCapabilitiesTest, AddFunctions) {
    EngineCapabilities caps;
    caps.add_function("add").add_function("subtract");
    
    EXPECT_TRUE(caps.supports_function("add"));
    EXPECT_FALSE(caps.supports_function("multiply"));
}

TEST(MockEngineTest, BasicInterface) {
    auto engine = make_engine<MockEngine>();
    
    auto info = engine->get_info();
    EXPECT_EQ(info.name, "MockEngine");
    
    auto caps = engine->get_capabilities();
    EXPECT_TRUE(caps.supports_relation("read"));
}

// Made with Bob
